packages/Menu/src/Menu.tsx

```diff
@@ -0,0 +1,173 @@
+   *
+   * default: `false`
+   */
+  active: boolean;
+
+  /**
+   * Determines the alignment of the Menu relative to a trigger element.
+   *
+   * default: `bottom`
+   */
+  align: Align;
+
+  /**
+   * Determines the justification of the Menu relative to a trigger element.
+   *
+   * default: `end`
+   */
+  justify: Justify;
+
+  /**
+   * Content that will appear inside of the Menu.
+   */
+  children?: React.ReactElement;
+
+  /**
+   * className applied to Menu.
+   */
+  className?: string;
+
+  /**
+   * A reference to the element against which the Menu will be positioned.
+   */
+  refEl?: React.RefObject<HTMLElement>;
+
+  /**
+   * A slot for the element used to trigger the Menu. Passing a trigger allows
+   * Menu to control opening and closing itself internally.
+   */
+  trigger?: React.ReactElement;
+
+  /**
+   * Specifies that the popover content will appear portaled to the end of the DOM,
+   * rather than in the React subset of the DOM tree.
+   *
+   * default: `true`
+   */
+  usePortal?: boolean;
+}
+
+/**
+ * # Menu
+ *
+ * Menu component
+ *
+ * ```
+<button>
+  <Menu active={true}>
+    <MenuItem>Hello World!</MenuItem>
+  </Menu>
+</button>
+ * ```
+ * @param props.children Content to appear inside of Menu.
+ * @param props.active Boolean to describe whether or not Menu is active.
+ * @param props.className Classname applied to Menu.
+ * @param props.align Alignment of Menu relative to another element: `top`, `bottom`, `left`, `right`.
+ * @param props.justify Justification of Menu relative to another element: `start`, `middle`, `end`.
+ * @param props.refEl Reference element that Menu should be positioned against.
+ * @param props.usePortal Boolean to describe if content should be portaled to end of DOM, or appear in DOM tree.
+ * @param props.trigger Trigger element to set active state of Menu, makes component controlled
+ *
+ */
+
+function Menu({
+  align = 'bottom',
+  justify = 'end',
+  usePortal = true,
+  active,
+  children,
+  className,
+  refEl,
+  trigger,
+  ...rest
+}: Props) {
+  const [isActive, setActiveState] = useState(false);
+
+  const syntheticToggleEventHandler: EventHandler<React.SyntheticEvent> = e => {
+    e.nativeEvent.stopImmediatePropagation();
+    setActiveState(!isActive);
+  };
+
+  const nativeToggleEventHandler = (e: Event) => {
+    e.stopImmediatePropagation();
+    setActiveState(!isActive);
+  };
+
+  const handleEscape = (e: KeyboardEvent) => {
+    e.keyCode === 27 && nativeToggleEventHandler(e);
+  };
+
+  const popoverContent = (
+    <Popover
+      key="popover"
+      active={trigger ? isActive : active}
+      align={align}
+      justify={justify}
+      refEl={refEl}
+      usePortal={usePortal}
+    >
+      <div {...rest} className={cx(rootMenuStyle, className)} role="menu">
+        {children}
+      </div>
+    </Popover>
+  );
+
+  let triggerElement: React.ReactNode = null;
+
+  if (trigger) {
+    useEffect(() => {
```

Comment:
So the reason that this is conditional is because we do not want to add any sort of activate/deactivate functionality if the component is uncontrolled. (If a trigger is provided, that means they're giving us internal state control, and therefore we want to activate the hook)

I think there's a way to provide a dependency array that watches to know when to re-run the effect, just have to think about what that would look like. I also think that it's possible to break a sub-component of the hook out into it's own custom hook (useKey and pass a whatever key to listen for). 

I don't, however, think this should be upstreamed into Popover. Popover is totally uncontrolled at the moment, so I don't think this belongs there.

Definitely would love to hear your thoughts on all of the above! You too @DesignerDave

Reply:
I could be wrong, but I don't think useEffect has that functionality – you may be thinking of `useMemo` or `useCallback`. Functionally, this should work just fine:
```js
useEffect(() => {
  if (trigger && isActive) {
    document.addEventListener('click', nativeToggleEventHandler, {
      once: true,
    });
    document.addEventListener('keydown', handleEscape, { once: true });
  }

  return () => {
    document.removeEventListener('click', nativeToggleEventHandler);
    document.removeEventListener('keydown', handleEscape);
  };
});

if (trigger) {
  triggerElement = React.cloneElement(trigger, {
    onClick: syntheticToggleEventHandler,
    children: [...trigger.props.children, popoverContent],
  });
}
```

On the custom hook, we could potentially break these listeners out, and have a dependency API for it. We're going to want to use this functionality in several places, so if ya can come up with a solid custom-hook implementation of this I'm all for it!

Also in total agreement on not upstreaming this functionality to Popover. It was an explicit decision to not have an uncontrolled version of it, and my opinion on that hasn't changed since. Popover is for controlling positioning, and visibility of the element, and not for maintaining control over its own active state.